EIPS/eip-X-hybrid-casper-ffg.md

```diff
@@ -0,0 +1,147 @@
+
+The FFG contract can be layered on top of any block proposal mechanism, providing finality to the underlying chain. This EIP proposes layering FFG on top of the existing PoW block proposal mechanism as a conservative step-wise approach in the transition to full PoS. The new FFG staking mechanism requires minimal changes to the protocol, allowing us to fully test and vet FFG on PoW before moving to a validator based block proposal mechanism.
+
+## Parameters
+
+* `HYBRID_CASPER_FORK_BLKNUM`: TBD
+* `CASPER_ADDR`: TBD
+* `CASPER_CODE`: see below
+* `CASPER_BALANCE`: 5e24 (5_000_000 eth)
+* `SIGHASHER_ADDR`: TBD
+* `SIGHASHER_CODE`: see below
+* `PURITY_CHECKER_ADDR`: TBD
+* `PURITY_CHECKER_CODE`: see below
+* `NULL_SENDER`: `2**160 - 1`
+* `NEW_BLOCK_REWARD`: 6e17 wei (0.6 ETH)
+* `NON_REVERT_MIN_DEPOSIT`: amount in wei decided by client
+
+### Casper Contract Parameters
+
+* `EPOCH_LENGTH`: 50 blocks
+* `WITHDRAWAL_DELAY`: 15000 epochs
+* `DYNASTY_LOGOUT_DELAY`: 700 dynasties
+* `SIGHASH_ADDR`: TBD
+* `PURITY_CHECKER_ADDR`: TBD
+* `BASE_INTEREST_FACTOR`: TBD
+* `BASE_PENALTY_FACTOR`: TBD
+* `MIN_DEPOSIT_SIZE`: 1000e18 wei (1000 ETH)
+
+## Specification
+
+### Deploying Casper Contract
+
+If `block.number == HYBRID_CASPER_FORK_BLKNUM`, then when processing the block, before processing any transactions:
+
+* set the code of `SIGHASHER_ADDR` to `SIGHASHER_CODE`
+* set the code of `PURITY_CHECKER_ADDR` to `PURITY_CHECKER_CODE`
+* set the code of `CASPER_ADDR` to `CASPER_CODE`
+* set balance of `CASPER_ADDR` to `CASPER_BALANCE` for issuance
+
+### Initialize Epochs
+
+If `block.number >= HYBRID_CASPER_FORK_BLKNUM and block.number % EPOCH_LENGTH == 0`, then execute a call with the following parameters at the start of the block:
+
+* `SENDER`: NULL_SENDER
+* `GAS`: 3141592
+* `TO`: CASPER_ADDR
+* `VALUE`: 0
+* `DATA`: <encoded call casper_translator.encode('initialize_epoch', [floor(block.number / EPOCH_LENGTH)])>
+
+This transaction utilizes no gas.
+
+### Casper Votes
+
+If `block.number >= HYBRID_CASPER_FORK_BLKNUM`, then:
+
+* all successful `vote` transactions to `CASPER_ADDR` with sender as `NULL_SENDER` must be included at the end of the block
+* all successful `vote` transactions to `CASPER_ADDR` with sender as `NULL_SENDER` utilize no gas
```

Comment:
also the "utilize no gas" needs clarification. The hackmd link dives a little further into it, but it can mean a few things:
  - these transactions can use unlimited gas and pay no fee
  - these transactions can use up to the remaining amount of gas in the block, with the gas counted towards the limit, but pay no fee
  - these transactions can use up to the remaining amount of gas in the block, with the gas not counted towards the limit, and pay no fee
  - these transactions can use up to the block gas limit, with the gas not counted towards the limit, and the fee is refunded after the call succeeds.
  - etc.

Without diving into the code of pyethereum right now, it's impossible to know which of the variants it is. In the hope of avoiding consensus failures, clarification would be appreciated.

Reply:
This PR was accidentally opened... but thanks for your comments!

Tightening up the spec now.
This spec will be a little different than the implementation guide. 
Draft PR coming for review shortly :)